### ISSUE TITLE
[no-floating-promises] False negative when linting across file boundaries

### ISSUE THREAD

**Issue by mwakizaka** [typescript-eslint/typescript-eslint#1688]
**Abstract**

I'm just a beginner of typescript. So I'm sorry if I might have something wrong.
Suppose there are 2 files(named `main.ts` and `lib.ts` respectively) deployed in single directory.
I ran eslint against 2 files and expected to see `no-floating-promises` errors, but I didn't.
Is it possible to check `no-floating-promises` across file boundaries? 

**Repro**

- rule
```JSON
{
  "rules": {
    "@typescript-eslint/no-floating-promises": "error"
  }
}
```

- main.ts
```TS(main.ts)
import lib = require('./lib');
import requestPromise = require("request-promise");

requestPromise({});
lib.getRequestPromise({});
```

- lib.ts
```TS(lib.ts)
import requestPromise = require("request-promise");

const getRequestPromise = async function(options) {
    return requestPromise(options);
};

module.exports = {
    getRequestPromise: getRequestPromise,
};
```

**Expected Result**
```
/***/lib.ts
  3:27  warning  Missing return type on function  @typescript-eslint/explicit-function-return-type
/***/main.ts
  4:1   error    Promises must be handled appropriately  @typescript-eslint/no-floating-promises
  5:1   error    Promises must be handled appropriately  @typescript-eslint/no-floating-promises
```

**Actual Result**
```
/***/lib.ts
  3:27  warning  Missing return type on function  @typescript-eslint/explicit-function-return-type
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.22.0` |
| `@typescript-eslint/parser`        | `2.22.0` |
| `TypeScript`                       | `3.8.3` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `12.13.1` |
| `npm`                              | `6.12.1` |

**Comment by bradzacher** [comment]
Could you provide a repro repo?
The rule works across module boundaries, but likely your things are getting typed incorrectly due to your setup.

A small, self contained, repro repo would make it easier to investigate.

**Comment by mwakizaka** [comment]
Hi, thank you for your response.
Here's a repo to reproduce. https://github.com/mwakizaka/no_floating_promise_trial
Please check it out.

**Comment by bradzacher** [comment]
Looking at everything, there's a lot wrong with it. The two things that are breaking the rule are:

- You don't have types for the `request-promise` package, so it's typed as `any`, and all functions you're calling on it are thusly typed as `any`.
- you're using `module.export` instead of just `export`ing the function. TS does not understand that syntax in modules, so it is typing your module as `any`.

Other, unrelated problems with the code:
- you're using `import foo = require('foo')` instead of just using proper import syntax, which typescript doesn't support when targeting ES2017 like your config is.
- you're not using the `strict` compiler option, so your code is missing types (i.e. your `options` argument has no type).

TL;DR - everything is untyped, so the rule can't tell it's promises.

**Comment by mwakizaka** [comment]
Thank you so much! I need to learn typescript more to type the codebase first.
